flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierProducerAgent.java

```diff
@@ -129,6 +129,7 @@ public boolean tryWrite(
         }
         if (finishedBuffer.isBuffer()) {
             memoryManager.transferBufferOwnership(bufferOwner, this, finishedBuffer);
+            memoryManager.ensureCapacity();
```

Comment:
How could this prevent hanging? IIUC, what `ensureCapacity` does is simply to request new buffers if the number of reserved buffers cannot be satisfied. However, such request may not be fulfilled, and we still may hang in `tryWrite`.

Reply:
Yes, we should ensure the capacity before starting a new segment instead of transferring ownership. I've change the implementation, please take a look.